### PR TITLE
Fix type resolving issue for intersection types having a type def for the same type

### LINKED ISSUE #35624: Incorrect type resolving for intersection types when there's a type def for the same type
Consider the following case:
```ballerina
type Foo record {|
    int a;
|};

type FooReadonly Foo & readonly;

function test() {
    Foo & readonly v; // invalid type symbol resolved for the intersection type here
}
```
If we take a look at the type symbol created for the intersection type above, it'll be the same type symbol created for `FooReadonly`. There shouldn't be a connection to that type def for other similar intersection types created. This leads to the wrong type symbol instance being returned to the Semantic API users for cases like this.

### PR DESCRIPTION
**Pull request by sanjana**
Fixes #35624

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

**Review by DamithaSenevirathne — DISMISSED** [review]
LGTM

**Comment by dulajdilshan** [comment]
Shall we re-sync with the `2201.2.x ` branch?

File: compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/SymbolFactory.java [hunk 1]
```diff
@@ -335,11 +339,31 @@ public BallerinaVariableSymbol createVariableSymbol(BVarSymbol symbol, String na
             symbolBuilder.withAnnotation(createAnnotationSymbol((BAnnotationAttachmentSymbol) annot));
         }
 
+        TypeSymbol typeDescriptor;
+        if (symbol.type.tag == TypeTags.INTERSECTION && symbol.type.tsymbol.getOrigin() == SymbolOrigin.VIRTUAL
+                || isReadonlyIntersectionArrayType(symbol.type)) {
+            typeDescriptor = typesFactory.getTypeDescriptor(symbol.type, symbol.type.tsymbol, true);
+        } else {
+            typeDescriptor = typesFactory.getTypeDescriptor(symbol.type);
```
Comment (dulajdilshan) [inline compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/SymbolFactory.java:347]: Let's try to include this logic inside the type-factory (if possible)

File: compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/SymbolFactory.java [hunk 2]
```diff
@@ -335,11 +339,31 @@ public BallerinaVariableSymbol createVariableSymbol(BVarSymbol symbol, String na
             symbolBuilder.withAnnotation(createAnnotationSymbol((BAnnotationAttachmentSymbol) annot));
         }
 
+        TypeSymbol typeDescriptor;
+        if (symbol.type.tag == TypeTags.INTERSECTION && symbol.type.tsymbol.getOrigin() == SymbolOrigin.VIRTUAL
```
[the remark this answers is not in the document]
Comment (sanjana) [reply compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/SymbolFactory.java:343]: Yes.

**Review by dulajdilshan — APPROVED** [review]
LGTM

**Pull request merged**